### PR TITLE
MAE/MLM for multimodal encoder pretraining, unimodal pretraining, finetuning eval

### PR DESCRIPTION
**Pull request by gmittal**
Lots of stuff

### CODE REVIEW

File: src/open_clip/flava_data.py [hunk 1]
```diff
@@ -28,23 +28,47 @@ def collate(example_list):
...
+
+def get_mlm_collate(hf_tokenizer, mlm_prob=0.15):
+    assert isinstance(hf_tokenizer, HFTokenizer), 'tokenizer must be HFTokenizer'
+    tokenizer = hf_tokenizer.tokenizer
+    mlm_collator = DataCollatorForLanguageModeling(
+        tokenizer=tokenizer,
+        mlm=True,
+        mlm_probability=mlm_prob,
+    )
+    pad_token_id = tokenizer.pad_token_id
```
Comment (apsdehal) [inline src/open_clip/flava_data.py]: Just a note: we should use same MLM strategy this time on both text and VL side. FLAVA used whole-word masking for VL but token masking for MLM.

File: src/open_clip/flava_data.py [hunk 2]
```diff
@@ -28,23 +28,47 @@ def collate(example_list):
...
 
         return batch
 
     return collate
+
+
+def get_mlm_collate(hf_tokenizer, mlm_prob=0.15):
+    assert isinstance(hf_tokenizer, HFTokenizer), 'tokenizer must be HFTokenizer'
+    tokenizer = hf_tokenizer.tokenizer
+    mlm_collator = DataCollatorForLanguageModeling(
```
Comment (apsdehal) [inline src/open_clip/flava_data.py:55]: Something to be aware about: `DataCollatorForLanguageModeling` updates the `input_ids` inplace which is something you might not be expecting in case you are passing them from somewhere else. Since this screw up contrastive loss as it will see masked text.

File: src/open_clip/loss.py [hunk 3]
```diff
@@ -152,6 +152,46 @@ def forward(self, itm_logits, itm_labels):
...
+    def __init__(self, norm_pix_loss):
+        super().__init__()
+        self.norm_pix_loss = norm_pix_loss
+
+    def patchify(self, imgs, p):
+        """
+        imgs: (N, 3, H, W)
+        x: (N, L, patch_size**2 *3)
+        """
+        assert imgs.shape[2] == imgs.shape[3] and imgs.shape[2] % p == 0
```
Comment (apsdehal) [inline src/open_clip/loss.py]: Usually with assert write what was the error.

File: src/open_clip/loss.py [hunk 4]
```diff
@@ -152,6 +152,46 @@ def forward(self, itm_logits, itm_labels):
...
+        """
+        patch_area = pred.shape[-1] / 3
+        patch_size = int(patch_area**0.5)
+        target = self.patchify(imgs, patch_size)
+        if self.norm_pix_loss:
+            mean = target.mean(dim=-1, keepdim=True)
+            var = target.var(dim=-1, keepdim=True)
+            target = (target - mean) / (var + 1.e-6)**.5
+
+        loss = (pred - target) ** 2
```
Comment (apsdehal) [inline src/open_clip/loss.py]: If you want this to be more efficient, you can possibly do `pred[mask] - target[mask]` as well.

File: src/training/data.py [hunk 5]
```diff
@@ -21,12 +21,38 @@
...
+        self.text_key = text_key
+        logging.debug('Done loading data.')
+
+        self.tokenize = tokenizer
+
+    def __len__(self):
+        return len(self.df)
+
+    def __getitem__(self, idx):
+        text = self.tokenize([str(self.df[idx][self.text_key])])[0]
```
Comment (apsdehal) [inline src/training/data.py:52]: If this is HFTokenizer, won't it return a dict with `input_ids` ?

File: src/training/data.py [hunk 6]
```diff
@@ -97,7 +123,7 @@ def get_dataset_size(shards):
     return total_size, num_shards
 
 
-def get_imagenet(args, preprocess_fns, split):
+def get_imagenet(args, preprocess_fns, split, flava_unimodal=False, collate_fn=None):
```
Comment (apsdehal) [inline src/training/data.py]: Why does the dataset have to know about the model's config?

Comment (apsdehal) [reply src/training/data.py]: (which basically tells something is not right in the design)

File: src/training/data.py [hunk 7]
```diff
@@ -133,11 +160,19 @@ def get_imagenet(args, preprocess_fns, split):
...
+    shuffle = flava_unimodal and sampler is None
+    batch_size = args.flava_unimodal_mae_batch_size if flava_unimodal else args.batch_size
+
     dataloader = torch.utils.data.DataLoader(
         dataset,
-        batch_size=args.batch_size,
+        batch_size=batch_size,
+        shuffle=shuffle,
         num_workers=args.workers,
+        pin_memory=flava_unimodal,
```
Comment (apsdehal) [inline src/training/data.py]: I would suggest using `pin_memory` and `drop_last` as parameters and then set them from caller based on `flava_unimodal`.

File: src/training/data.py [hunk 8]
```diff
@@ -133,11 +160,19 @@ def get_imagenet(args, preprocess_fns, split):
     else:
         sampler = None
 
+    sampler = DistributedSampler(dataset) if args.distributed and flava_unimodal else None
+    shuffle = flava_unimodal and sampler is None
+    batch_size = args.flava_unimodal_mae_batch_size if flava_unimodal else args.batch_size
```
Comment (apsdehal) [inline src/training/data.py]: Same suggestion as below for this. Take this as high level parameter.

File: src/training/data.py [hunk 9]
```diff
@@ -448,6 +483,36 @@ def get_csv_dataset(args, preprocess_fn, is_train, epoch=0, tokenizer=None, coll
...
+
+    dataloader = DataLoader(
+        dataset,
+        batch_size=args.flava_unimodal_mlm_batch_size,
+        shuffle=shuffle,
+        num_workers=args.workers,
+        pin_memory=True,
+        sampler=sampler,
+        collate_fn=collate_fn,
+        drop_last=True,
```
Comment (apsdehal) [inline src/training/data.py]: Hmm. Why always true?

File: src/training/data.py [hunk 10]
```diff
@@ -448,6 +483,36 @@ def get_csv_dataset(args, preprocess_fn, is_train, epoch=0, tokenizer=None, coll
...
+    num_samples = len(dataset)
+    sampler = DistributedSampler(dataset) if args.distributed else None
+    shuffle = sampler is None
+
+    dataloader = DataLoader(
+        dataset,
+        batch_size=args.flava_unimodal_mlm_batch_size,
+        shuffle=shuffle,
+        num_workers=args.workers,
+        pin_memory=True,
```
Comment (apsdehal) [inline src/training/data.py]: Same why always true? `pin_memory` usage can be some times tricky as well.

File: src/training/data.py [hunk 11]
```diff
@@ -448,6 +483,36 @@ def get_csv_dataset(args, preprocess_fn, is_train, epoch=0, tokenizer=None, coll
...
+        sampler=sampler,
+        collate_fn=collate_fn,
+        drop_last=True,
+    )
+    dataloader.num_samples = num_samples
+    dataloader.num_batches = len(dataloader)
```
Comment (apsdehal) [inline src/training/data.py:555]: I don't like that we are adding attributes on the fly to the dataloader.

File: src/training/finetune_glue.py [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+        )
+    model.train()
+    metrics = metric.compute()
+    metrics["loss"] = val_loss / samples_seen
+    return metrics
+
+
+def train_one_epoch(model, data, epoch, optimizer, scheduler, early_stop, device, args):
+    model.train()
+    progress_bar = tqdm(total=len(data["train"]))
```
Comment (apsdehal) [inline src/training/finetune_glue.py]: When you are using distributed data sampler. You also need to call `set_epoch` on the dataloader so as to make sure each epoch is loading data in different order.

File: src/training/finetune_hm.py [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+        }
+
+
+class FLAVAMultimodalClassifier(nn.Module):
+    def __init__(self, encoder, embed_dim, num_labels):
+        super().__init__()
+
+        self.encoder = encoder
+        self.logits_proj = nn.Sequential(
+            nn.Linear(embed_dim, 1536),
```
Comment (apsdehal) [inline src/training/finetune_hm.py]: Probably don't need to hardcode?

File: src/open_clip/flava_model.py [hunk 14]
```diff
@@ -80,19 +168,21 @@ def forward(self, x: torch.Tensor, attn_mask: Optional[torch.Tensor] = None):
...
 
-    timm_model_name: str = None  # a valid model name overrides layers, width, patch_size
-    timm_model_pretrained: bool = False  # use (imagenet) pretrained weights for named model
-    timm_pool: str = 'avg'  # feature pooling for timm model ('abs_attn', 'rot_attn', 'avg', '')
-    timm_proj: str = 'linear'  # linear projection for timm model output ('linear', 'mlp', '')
-    timm_proj_bias: bool = False  # enable bias final projection
+    # MAE parameters
+    mae_mask_ratio: float = 0.75
+    mae_decoder_layers: int = 2,
+    mae_decoder_width: int = 512,
```
Comment (apsdehal) [inline src/open_clip/flava_model.py]: Why do they have `,` at the end?

File: src/open_clip/flava_model.py [hunk 15]
```diff
@@ -127,164 +294,258 @@ def __init__(
...
+        self.mm_patch_mask_token = nn.Parameter(multimodal_cfg.width ** -0.5 * torch.randn(multimodal_cfg.width))
+        self.mm_mae_head = nn.Sequential(
+            nn.Linear(embed_dim, embed_dim * 4),
+            nn.GELU(),
+            nn.Linear(embed_dim * 4, vision_cfg.patch_size ** 2 * 3, bias=True),  # patch reconstruction
+        )
+
+        # ITM
+        self.itm_head = nn.Sequential(
+            nn.Linear(embed_dim, embed_dim * 4),
```
Comment (apsdehal) [inline src/open_clip/flava_model.py]: Where did this `* 4` number came from? Are you taking inspiration from transformer's MLP block?

File: src/open_clip/flava_data.py (hunk 2 above)
Comment (gmittal) [reply src/open_clip/flava_data.py:55]: Good catch! I checked that this does not cause any issues and that the original `text` is unharmed. Also verified this by looking at the MLM loss which looks fine.

File: src/training/data.py (hunk 5 above)
Comment (gmittal) [reply src/training/data.py:52]: It uses a wrapper around the HF tokenizer written by the OpenCLIP team: https://github.com/gmittal/open_clip/blob/8078267fc9385a8dfada34e24496cde8ec3ff430/src/open_clip/tokenizer.py#L188-L202

File: src/training/data.py (hunk 11 above)
Comment (gmittal) [reply src/training/data.py:555]: I don't like it either, but it seems to be a thing that the OpenCLIP repo does a good amount (and the trainer expects it).

File: src/open_clip/flava_model.py (hunk 15 above)
Comment (gmittal) [reply src/open_clip/flava_model.py]: yes

File: src/open_clip/flava_data.py (hunk 1 above)
Comment (gmittal) [reply src/open_clip/flava_data.py]: Yep, token-level masking is used for both MLM and VL right now.
